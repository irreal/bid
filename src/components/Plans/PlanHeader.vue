<template>
  <v-layout align-end py-3>
    <span class="headline">{{ plan.title }}</span>
    <span>Ispunjenost plana: {{ Math.round(plan.percent_complete) }}%</span>
    <span>Autor plana: {{ plan.created_by }}</span>
  </v-layout>
</template>
<script>
export default {
  name: "PlanHeader",
  props: {
    plan: {
      required: true,
      type: Object
    }
  },
  computed: {}
};
</script>
<style scoped lang="scss"></style>
