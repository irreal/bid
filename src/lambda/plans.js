const mongoose = require("mongoose");
const planSchema = new mongoose.Schema({
  title: String,
  period_start: Date,
  period_end: Date,
  date_added: Date,
  percent_complete: Number,
  created_by: String,
  allowed_roles: [String],
  allowed_users: [String]
});
var Plan = mongoose.model("plans", planSchema);
export async function handler(event, context, callback) {
  const { user } = context.clientContext;
  if (!user) {
    callback(null, {
      statusCode: 401,
      body: "missing client context"
    });
    return;
  }
  try {
    const uri = process.env.MONGO_CONNECTION;
    await mongoose.connect(uri, {
      useNewUrlParser: true
    });

    if (event.httpMethod === "POST") {
      try {
        const params = JSON.parse(event.body);
        console.log("params: ", params);
        if (!params.title) {
          throw new Error("missing parameter title");
        }
        if (!params.period_start) {
          throw new Error("missing parameter period start");
        }
        if (!params.period_end) {
          throw new Error("missing parameter period end");
        }

        var newPlan = new Plan({
          title: params.title,
          percent_complete: 0,
          period_start: params.period_start,
          period_end: params.period_end,
          date_added: new Date(),
          created_by: user.email
        });
        await newPlan.save();
        callback(null, {
          statusCode: 201,
          body: '"ok"'
        });
      } catch (err) {
        console.log("evo gre≈°ke, ", err);
        callback(null, {
          statusCode: 400,
          body: `"Invalid request\n${err.toString()}"`
        });
      }
      return;
    }
    if (event.httpMethod !== "GET") {
      callback(null, {
        statusCode: 405
      });
      return;
    }

    let filter = {};
    if (!user.app_metadata.roles.includes("admin")) {
      filter = {
        $or: [
          {
            created_by: user.email
          },
          {
            allowed_roles: {
              $in: user.app_metadata.roles
            }
          },
          {
            allowed_users: user.email
          }
        ]
      };
    }
    const results = await Plan.find(
      filter,
      "_id title percent_complete period_start period_end created_by"
    );
    await mongoose.connection.close();
    callback(null, {
      statusCode: 200,
      body: JSON.stringify(results)
    });
  } catch (err) {
    await mongoose.connection.close();
    console.log("error accessing plans ", err);
    callback(err);
  }
}
