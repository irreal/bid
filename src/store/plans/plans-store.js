import Axios from "axios";

export default {
  namespaced: true,
  state: { items: [] },
  mutations: {
    set: (state, plans) => {
      state.items = plans;
    }
  },
  getters: {
    all: state => state.items,
    find: state => id => state.items.find(i => i._id === id)
  },
  actions: {
    load: ({ commit }) => {
      Axios.get("/.netlify/functions/plans").then(resp => {
        commit("set", resp.data);
      });
    }
  }
};
