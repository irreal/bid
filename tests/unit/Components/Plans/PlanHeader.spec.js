import { shallowMount, createLocalVue } from "@vue/test-utils";
import PlanHeader from "@/components/Plans/PlanHeader";
import Vuetify from "vuetify";
import Vue from "vue";
import plans from "../../fixtures/plans";

const localVue = createLocalVue();
Vue.use(Vuetify);

describe("Plan Header component", () => {
  const build = () => {
    const plan = shallowMount(PlanHeader, {
      localVue,
      propsData: {
        plan: plans[0]
      }
    });

    return {
      plan
    };
  };

  it("renders", () => {
    // arrange
    const { plan } = build();
    // assert
    expect(plan.html()).toMatchSnapshot();
  });
  it("displays plan title", () => {
    const { plan } = build();
    expect(plan.html()).toContain(plans[0].title);
  });

  it("displays percentage complete", () => {
    const { plan } = build();
    expect(plan.html()).toContain(Math.round(plans[0].percent_complete) + "%");
  });
  it("displays plan creator", () => {
    const { plan } = build();
    expect(plan.html()).toContain(plans[0].created_by);
  });
});
