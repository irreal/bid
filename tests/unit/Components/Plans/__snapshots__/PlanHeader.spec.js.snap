// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Plan Header component renders 1`] = `<v-layout-stub tag="div" align-end="" py-3=""><span class="headline">moj plan novi</span> <span>Ispunjenost plana: 2%</span> <span>Autor plana: milos.s.pfc@gmail.com</span></v-layout-stub>`;
